### PR TITLE
[GrandCentral] Ignore deduped Companion views

### PR DESCRIPTION
**Pull request by prithayan**
This commit ensures that, if a companion module has two exactly same view annotations, then only one interface is generated.
This can occur when companions are deduped, and the corresponding View annotations are added as nonlocal after dedup.
If all the elements of the interface are exactly the same, then the companion must generate only one of the interfaces.
This commit still ensures that only a single instance of the companion exists. If multiple instances of the companion exist then GrandCentral will fail. This only handles the case when the parent module instantiating the companion is also deduped, thus ensuring only a single instance of the companion exists.

### CODE REVIEW

**Comment by seldridge** [comment]
I think this will work. If we got this route, this should also error if all NLA interfaces are not the same.

It is likely better to implement the `FExtModuleOp` plan directly. I.e., change the representation of the interface to be an `FExtModuleOp` which is specially annotated to indicate that it is an interface and should be lowered as such. This would do the same thing, except block dedup if things go awry.

**Comment by prithayan** [comment]
@seldridge, Even if we use `FExtModuleOp`, we still have to do the per element comparison of the `AugmentedBundleType` elements in the GrandCentral pass right ?

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2103]: Can the PR include a `TODO` that this is something that we want to change in the future?

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2121]: Nit:

```suggestion
        // Now compare the existing interface with this.
```

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2129]: How does the `index()` control whether or not this is a logic or a sub-interface?

File: test/Dialect/FIRRTL/grand-central.mlir
Comment (seldridge) [inline test/Dialect/FIRRTL/grand-central.mlir:1158]: Can the PR use `CHECK-NEXT` vs. `CHECK` when appropriate? 

Nits: 

1. Can the PR align everything under the firrtl.module so, when viewed it looks like:

```
firrtl.module
  ...
  ...
```

2. The PR should also include `CHECK-NOT` to make sure that sub-interfaces don't get created, yes?

**Review by seldridge — COMMENTED** [review]
Generally looks good. Some nits/questions. The main thing is the test probably wants to include more `CHECK-NOT` that extra sub-interfaces aren't created.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp]: Added the TODO.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp]: This was not handling sub-interfaces correctly. Updated the implementation to just record the interface signal ops.

**Comment by prithayan** [comment]
Thanks @seldridge for the review, I have updated this to check that each `sv.interface.signal` ops match, that should ensure that sub-interfaces are matched properly. Please review this once again.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:1869]: `hasMember<FlatSymbolRefAttr>("circt.nonlocal")` should work!

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2066]: Mega supreme nit: 
```suggestion
    // If it's a verbatim op, no need to check the string, because the interface
    // names might not match. As long as the signal ops match that is
    // sufficient.
```

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2068]: This check _could_ be erroneously true. To actually fix this would require a pretty big refactor of how `traverseBundle` works. Can you add a TODO and open a tracking issue once this lands?

File: test/Dialect/FIRRTL/grand-central.mlir
Comment (seldridge) [inline test/Dialect/FIRRTL/grand-central.mlir:1210]: Mega-nit:
```suggestion
  // CHECK:      sv.interface @[[VectorOfBundleView]] attributes
  // CHECK-NOT:    sv.interface @VectorOfBundleView_0 
  // CHECK:      sv.interface @Bundle2
```

**Review by seldridge — APPROVED** [review]
This looks fine for the approach of allowing this.

I'd like to have explicit sign-off from @darthscsi that he is okay with this approach before landing in light of the discussion that we really should be disallowing deduplication of input port reference types (upwards XMRs).

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:571]: This is a major code smell.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2138]: Can't you fix this up when you first add the companion when you find it (much above)?

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2113]: The same builder could be re-used by moving this outside the loop and using `InsertionGuard` or just being careful.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2194]: The PR could add an assert that `topIface` is definitely set at this point.

**Review by seldridge — APPROVED** [review]
Looks good!

Thanks for reworking this to not create a bunch of throwaway interfaces. 👍

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:1837]: I couldn't find the `Annotation::hasMember` , I recall something similar existing, was it removed ?

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:548]: The `emplace_back`, requires this explicit constructor, otherwise it's not able to find the appropriate one.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp]: Moved it out.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:2195]: Added a loop continue, if the interface elements is empty, that ensures, it is always initialized.

**Comment by prithayan** [comment]
Thanks for the review @seldridge and @darthscsi , merged this after refactoring the implementation.

File: lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:1475]: This will cause memory corruption when interfaceBuilder reallocates memory.

Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:1475]: Thanks @uenoku  for catching this. Fix in https://github.com/llvm/circt/pull/4351

**Pull request merged**
